### PR TITLE
Add option to draw easymotion on top

### LINKED ISSUE #276: EasyMotion shifting text
My favorite things about easy motion is that you don't need to move your gaze away from where you want to go. The text decorations seem to break that though since they get placed behind the text and shift everything on the line over. When there are a lot of matches on a given line, it can move the characters significantly.

Is there any way to avoid replacing the text in the buffer like the original easymotion did while also keeping the text in it's original positions?

### PR DESCRIPTION
**Pull request by jhgarner**
The original easymotion extension drew text directly on top of the
target characters. The fork drew text behind the target characters.
This commit adds an option to pick whichever you want.
Even if you pick on top of the target, you won't end up with broken
linting.

closes #276

### CODE REVIEW

**Comment by jhgarner** [comment]
I think there are some inconsistencies with motions at the ends of lines. I'll try to fix those and give a comment when that's done.

**Comment by jhgarner** [comment]
I think this is finally working with almost perfect accuracy. The code is a bit more complicated, but it makes using the plugin significantly nicer. If I don't catch any problems in the next few days, I'll remove the WIP tag.

**Comment by theol0403** [comment]
It seems to work great. One thing I noticed is that the text overlay is *slightly* too far to the left. 

image
image

This causes the shadow of the gutter to occasionally overlap the text:
image

**Comment by jhgarner** [comment]
I wonder if that has to do with the magic number of 8 that's used for the margin and width. Maybe there's some way to dynamically pick the right thing? I'll look into that.

**Comment by jhgarner** [comment]
@theol0403 I've switched the units from px to ch. I didn't see the problem on my machine so let me know if that fixed it for you.

**Comment by theol0403** [comment]
Yep, that fixed it. It now works beautifully on my machine :+1:

**Comment by jhgarner** [comment]
I've removed the WIP from the title. It seems to work pretty well.

**Comment by Shatur95** [comment]
The feature I missing. Offsetting lines makes easymotion very awkward to use. But I see that @asvetliakov is busy now. Maybe it would be better to let some right to the contributors, such as @jhgarner?

**Comment by asvetliakov** [comment]
Will look/review the PR within next week

**Comment by Shatur95** [comment]
@jhgarner, could you resolve merge conflicts?

File: package.json [hunk 1]
```diff
@@ -220,6 +220,11 @@
 					"title": "Unknown Highlight Groups",
 					"description": "Highlight Configuration for non defined group. Can accept 'vim' to use colors coming from vim or text decoration properties object",
 					"default": "vim"
+				},
+				"vscode-neovim.easymotionOnTop": {
```
Comment (asvetliakov) [inline package.json:224]: let's rename to `textDecorationsAtTop`

File: package.json [hunk 2]
```diff
@@ -1020,6 +1025,7 @@
 		"prettier": "^2.0.5",
 		"source-map-support": "^0.5.16",
 		"typescript": "^3.7.4",
+		"vsce": "^1.79.3",
```
Comment (asvetliakov) [inline package.json:1028]: not needed here

File: src/controller.ts [hunk 3]
```diff
@@ -1910,26 +1910,65 @@ export class NVIMPluginController implements vscode.Disposable {
         }
```
Comment (asvetliakov) [inline src/controller.ts:1910]: This should be in `highlight_manager.ts`

File: src/controller.ts [hunk 4]
```diff
@@ -1910,26 +1910,65 @@ export class NVIMPluginController implements vscode.Disposable {
...
-                    // const col = colNum - 1;
-                    const col = Utils.convertByteNumToCharNum(line, colNum - 1);
-                    const opt: vscode.DecorationOptions = {
-                        range: new vscode.Range(lineNum, col, lineNum, col),
-                        renderOptions: {
-                            before: {
-                                ...conf,
-                                ...conf.before,
-                                contentText: text,
+                    if (Utils.getEasymotionOnTop()) {
```
Comment (asvetliakov) [inline src/controller.ts:1921]: Pass an option to HL manager constructor. Getting workspace configuration every time is costly

File: src/controller.ts [hunk 5]
```diff
@@ -1910,26 +1910,65 @@ export class NVIMPluginController implements vscode.Disposable {
...
+                        // 8 seems to be a magic number. It works for me, but I'm not sure why...
+                        const width = text.length;
+                        // vim sends column in bytes, need to convert to characters
+                        const col = Utils.convertByteNumToCharNumOnTop(line, colNum);
```
Comment (asvetliakov) [inline src/controller.ts:1925]: No need for another function - `convertByteNumToCharNum()` is used now only for decorations. Also check how does it work with tabs

File: src/controller.ts [hunk 6]
```diff
@@ -1910,26 +1910,65 @@ export class NVIMPluginController implements vscode.Disposable {
...
-                    const col = Utils.convertByteNumToCharNum(line, colNum - 1);
-                    const opt: vscode.DecorationOptions = {
-                        range: new vscode.Range(lineNum, col, lineNum, col),
-                        renderOptions: {
-                            before: {
-                                ...conf,
-                                ...conf.before,
-                                contentText: text,
+                    if (Utils.getEasymotionOnTop()) {
+                        // 8 seems to be a magic number. It works for me, but I'm not sure why...
```
Comment (asvetliakov) [inline src/controller.ts:1922]: Not sure i understand this comment

File: src/controller.ts [hunk 7]
```diff
@@ -1910,26 +1910,65 @@ export class NVIMPluginController implements vscode.Disposable {
...
+                            );
+                        } else {
+                            const opt: vscode.DecorationOptions = {
+                                range: new vscode.Range(lineNum, col + text.length - 1, lineNum, col + text.length - 1),
+                                renderOptions: {
+                                    // Inspired by https://github.com/VSCodeVim/Vim/blob/badecf1b7ecd239e3ed58720245b6f4a74e439b7/src/actions/plugins/easymotion/easymotion.ts#L64
+                                    after: {
+                                        // If we try to draw off the end of the screen, VSCode will move the tet to the left.
+                                        // Each move VSCode makes to the left for us is one more we don't want to do.
+                                        margin: `0 0 0 -${Math.min(
```
Comment (asvetliakov) [inline src/controller.ts:1944]: What will happen if we won't use a negative right margin? Mind a share a screenshot?

File: src/utils.ts [hunk 8]
```diff
@@ -277,6 +277,19 @@ export function convertByteNumToCharNum(line: string, col: number): number {
     return currCharNum;
 }
 
+export function convertByteNumToCharNumOnTop(line: string, col: number): number {
+    let totalBytes = 0;
+    let currCharNum = 0;
+    while (totalBytes < col) {
+        if (currCharNum >= line.length) {
+            return currCharNum + (col - totalBytes);
```
Comment (asvetliakov) [inline src/utils.ts:285]: Why this is needed? Anyway, better to modify the existing function rather than creating a copy with slighly changed behavior

**Review by asvetliakov — CHANGES_REQUESTED** [review]

**Comment by Shatur95** [comment]
Maybe it would be easier to just open new PR since master have a lot of changes?

**Comment by asvetliakov** [comment]
>Maybe it would be easier to just open new PR since master have a lot of changes?

Agree

**Comment by jhgarner** [comment]
Thanks for the review. I'll close this issue, refactor it for the new master, and fix the comments you brought up.

**Comment by jrolfs** [comment]
↓ Can't wait to get rid of this! ↓
vscode-neovim--absolute-text-decorations


If anyone's looking for a workaround in the meantime though...

```json
  "customizeUI.stylesheet": {
    ".monaco-editor span[class*=' ced-1-TextEditorDecorationType']:not([class*=' ced-1-TextEditorDecorationType1'])::before": "position: absolute;"
  }
```

**Pull request closed without merging**